### PR TITLE
[Enhancement] New Callback event, before and after backward

### PR DESCRIPTION
**Pull request by muellerzr**
This PR introduces a new Callback event: `after_backward`. This is designed so custom backward pass implementations can be used (such as accelerate).

`after_backwards` is the same as `before_step`, but I opted to keep both because there isn't a clean way to combine the two into one without introducing too many changes. 

If a callback implements their own backward pass, they just need to attach a `RaiseBackwardException` in the implementation.

cc @jph00

### CODE REVIEW

**Comment by muellerzr** [comment]
ci failed unrelated to the current notebook

File: nbs/13a_learner.ipynb [hunk 1]
```diff
@@ -24,7 +24,20 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/opt/conda/lib/python3.9/site-packages/torch/cuda/__init__.py:145: UserWarning: \n",
```
Comment (jph00) [inline nbs/13a_learner.ipynb:32]: Can you fix this? It'll mess up the docs

File: nbs/13a_learner.ipynb [hunk 2]
```diff
@@ -723,7 +735,32 @@
    "cell_type": "code",
    "execution_count": null,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "ename": "RuntimeError",
```
Comment (jph00) [inline nbs/13a_learner.ipynb:740]: ...and this

**Review by jph00 — CHANGES_REQUESTED** [review]
Looks like you need to re-run this after fixing your cuda env

File: nbs/13a_learner.ipynb (hunk 1 above)
Comment (muellerzr) [reply nbs/13a_learner.ipynb:32]: Ack, yup sorry! Will do so momentarily

**Comment by warner-benjamin** [comment]
Could you update the Callback documentation to mention `after_backward` and what it is used for? Currently it reads:
```
- `before_backward`: called after the loss has been computed, but only in training mode (i.e. when the backward pass will be used)
- `before_step`: called after the backward pass, but before the update of the parameters. It can be used to do any change to the gradients before said update (gradient clipping for instance).
```

**Comment by muellerzr** [comment]
@warner-benjamin `after_backward` is synonymous with `before_step`. But everyone should use `before_step` really, hence why it was left unmentioned.

**Comment by warner-benjamin** [comment]
That strikes me as an important thing to mention.

**Pull request merged**